TELCODOCS-1820: kmod changes (#74713)
openshift/openshift-docs: modules/kmm-signing-kmods-in-a-prebuilt-image.adoc

// Module included in the following assemblies:
//
// * hardware_enablement/kmm-kernel-module-management.adoc

:_mod-docs-content-type: PROCEDURE
[id="kmm-signing-kmods-in-a-prebuilt-image_{context}"]
= Signing kmods in a pre-built image

Use this procedure if you have a pre-built image, such as an image either distributed by a hardware vendor or built elsewhere.

> The following YAML file adds the public/private key-pair as secrets with the required key names - `key` for the private key, `cert` for the public key. The cluster then pulls down the `unsignedImage` image, opens it, signs the kernel modules listed in `filesToSign`, adds them back, and pushes the resulting image as `containerImage`.

Reviewer: This is a soft suggestion if you're up for it, but this second sentence has enough actions in it where maybe putting these actions in a list would make for easier reading. But up to you, it's not too bad.

Author: No change.

> KMM should then load the signed kmods onto all the nodes with that match the selector. The kmods should be successfully loaded on any nodes that have the public key in their MOK database, and any nodes that are not secure-boot enabled, which will ignore the signature.

Reviewer: This is my opinion (sort of supported by the [ISG word usage entry for "should"](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#s)), but I think if the things here that "should" happen are the expected behavior, this can be written more directly and assertively:
```suggestion
KMM then loads the signed kmods onto all the nodes with that match the selector. The kmods are successfully loaded on any nodes that have the public key in their MOK database, and any nodes that are not secure-boot enabled, which will ignore the signature.
```

Author: Changed.

.Prerequisites

> * The `keySecret` and `certSecret` secrets have been created.

Reviewer 2: Same here, specify that the secrets should be in the same namespace.

Author: Changed to "The keySecret and certSecret secrets have been created in the same namespace as the rest of the resources."

.Procedure

> . Apply the YAML file:

Reviewer: Make this a bullet point since this is the only step in the procedure:
```suggestion
* Apply the YAML file:
```

Author: Changed.

+
[source,yaml]
----
---
apiVersion: kmm.sigs.x-k8s.io/v1beta1
kind: Module
metadata:
  name: example-module
spec:
  moduleLoader:
    serviceAccountName: default
    container:
      modprobe: <1>

>         moduleName: '<your module name>'

Reviewer: Same suggestion as earlier to make these values concise:
```suggestion
        moduleName: '<module_name>'
```
(replacing the spaces in the value with an underscore, though, is a requirement per our guidelines)

Author: Changed.

      kernelMappings:
        # the kmods will be deployed on all nodes in the cluster with a kernel that matches the regexp
        - regexp: '^.*\.x86_64$'
          # the container to produce containing the signed kmods

>           containerImage: <image name e.g. quay.io/myuser/my-driver:<kernelversion>-signed>

Reviewer: Same comment as above about concise and correctly formatted input values:
```suggestion
          containerImage: <image_name>
```

Since there's the example you want to show here, this is a good place to put another callout label and put the example in the description.

          sign:
            # the image containing the unsigned kmods (we need this because we are not building the kmods within the cluster)

>             unsignedImage: <image name e.g. quay.io/myuser/my-driver:<kernelversion> >

Reviewer: Same comment as above:
```suggestion
            unsignedImage: <image_name>
```

Author: Made these changes.

            keySecret: # a secret holding the private secureboot key with the key 'key'

>               name: <private key secret name>

Reviewer: Same comment about formatting:
```suggestion
              name: <private_key_secret_name>
```

Author: Changed.

            certSecret: # a secret holding the public secureboot key with the key 'cert'

>               name: <certificate secret name>

Reviewer: Same comment about formatting:
```suggestion
              name: <certificate_secret_name>
```

Author: Changed.

            filesToSign: # full path within the unsignedImage container to the kmod(s) to sign
              - /opt/lib/modules/4.18.0-348.2.1.el8_5.x86_64/kmm_ci_a.ko
  imageRepoSecret:
    # the name of a secret containing credentials to pull unsignedImage and push containerImage to the registry
    name: repo-pull-secret
  selector:
    kubernetes.io/arch: amd64
----

> <1> `modprobe` - The name of the kmod to load.

Reviewer: Same suggestion as earlier - seems like it's technically allowed, but I don't think the field name is necessary to repeat and it doesn't follow the examples in our guidelines:
```suggestion
<1> The name of the kmod to load.
```

Author: Changed.
